cps/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinition.java

```diff
@@ -0,0 +1,139 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.jenkinsci.plugins.workflow.cps;
+
+import hudson.Extension;
+import hudson.FilePath;
+import hudson.model.Action;
+import hudson.model.Computer;
+import hudson.model.Job;
+import hudson.model.Queue;
+import hudson.model.Run;
+import hudson.model.TaskListener;
+import hudson.model.TopLevelItem;
+import hudson.scm.SCM;
+import hudson.scm.SCMDescriptor;
+import hudson.slaves.WorkspaceList;
+import java.io.IOException;
+import java.util.Collection;
+import java.util.List;
+import jenkins.model.Jenkins;
+import org.jenkinsci.plugins.workflow.cps.persistence.PersistIn;
+import static org.jenkinsci.plugins.workflow.cps.persistence.PersistenceContext.JOB;
+import org.jenkinsci.plugins.workflow.flow.FlowDefinition;
+import org.jenkinsci.plugins.workflow.flow.FlowDefinitionDescriptor;
+import org.jenkinsci.plugins.workflow.flow.FlowExecutionOwner;
+import org.jenkinsci.plugins.workflow.steps.scm.GenericSCMStep;
+import org.jenkinsci.plugins.workflow.steps.scm.SCMStep;
+import org.jenkinsci.plugins.workflow.support.actions.WorkspaceActionImpl;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.Stapler;
+import org.kohsuke.stapler.StaplerRequest;
+
+@PersistIn(JOB)
+public class CpsScmFlowDefinition extends FlowDefinition {
+
+    private final SCM scm;
+    private final String scriptPath;
+
+    @DataBoundConstructor public CpsScmFlowDefinition(SCM scm, String scriptPath) {
+        this.scm = scm;
+        this.scriptPath = scriptPath;
+    }
+
+    public SCM getScm() {
+        return scm;
+    }
+
+    public String getScriptPath() {
+        return scriptPath;
+    }
+
+    @Override public CpsFlowExecution create(FlowExecutionOwner owner, TaskListener listener, List<? extends Action> actions) throws Exception {
+        for (Action a : actions) {
+            if (a instanceof CpsFlowFactoryAction2) {
+                return ((CpsFlowFactoryAction2) a).create(this, owner, actions);
+            }
+        }
+        FilePath dir;
+        Queue.Executable _build = owner.getExecutable();
+        if (!(_build instanceof Run)) {
+            throw new IOException("can only check out SCM into a Run");
+        }
+        Run<?,?> build = (Run<?,?>) _build;
+        Jenkins jenkins = Jenkins.getInstance();
```

Comment:
Right, there is no particular reason the same procedure could not be done on a slave; you would just need to allow a label expression to be configured. I think it can be skipped unless someone requests it.